### PR TITLE
Feature/mmdetection2.0

### PR DESCRIPTION
**Pull request by Ilya-Krylov**

### CODE REVIEW

File: pytorch_toolkit/object_detection/face-detection/face-detection-0200/config.py [hunk 1]
```diff
@@ -128,7 +132,7 @@
 checkpoint_config = dict(interval=1)
 # yapf:disable
 log_config = dict(
-    interval=100,
+    interval=10,
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/face-detection/face-detection-0200/config.py:135]: What's the reason of this change?

File: pytorch_toolkit/object_detection/face-detection/readme.md [hunk 2]
```diff
@@ -2,15 +2,13 @@
...
-| face-detection-0105 | 2.94 | 3.72 | 20.9 | 93.335 | 91.607 | 85.94 | 52.925 | [snapshot](https://download.01.org/opencv/openvino_training_extensions/models/object_detection/face-detection-0105.pth), [configuration file](./face-detection-0105/config.py) | 4 | 
-| face-detection-0106 | 340.06 | 65.84 | 34.6 | 94.666 | 94.506 | 93.203 | 83.095 | [snapshot](https://download.01.org/opencv/openvino_training_extensions/models/object_detection/face-detection-0106.pth), [configuration file](./face-detection-0106/config.py) | 8 | 
-| face-detection-0107 | 1.04 | 2.31 | 16.9 | 89.55 | 85.128 | 76.629 | 42.867 | [snapshot](https://download.01.org/opencv/openvino_training_extensions/models/object_detection/face-detection-0107.pth), [configuration file](./face-detection-0107/config.py) | 4 | 
-| face-detection-0108 | 2.5 | 2.31 | 20.8 | 92.251 | 89.683 | 83.888 | 50.476 | [snapshot](https://download.01.org/opencv/openvino_training_extensions/models/object_detection/face-detection-0108.pth), [configuration file](./face-detection-0108/config.py) | 4 |
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/face-detection/readme.md:13]: What about the last two?

File: pytorch_toolkit/object_detection/tests/export_tests_face_detection.py [hunk 3]
```diff
@@ -29,29 +29,21 @@ def setUp(self):
     return ExportTestCase
 
 
-class FaceDetection0100TestCase(face_detection_test_case('face-detection-0100', True)):
-    """ Test case for face-detection-0100 model export. """
+class FaceDetection0100TestCase(face_detection_test_case('face-detection-0200', True)):
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/tests/export_tests_face_detection.py:32]: Test cases use old model naming scheme. Is it by design?

File: pytorch_toolkit/object_detection/tests/export_tests_person_vehicle_bike_detection.py [hunk 4]
```diff
@@ -13,12 +13,12 @@
...
-    """ Test case for person-vehicle-bike-detection-crossroad-1016 export. """
+# from common.test_case import export_test_case
+#
+#
+# class PersonVehicleBikeDetectionCrossroad1016TestCase(
+#         export_test_case('person-vehicle-bike-detection',
+#                          'person-vehicle-bike-detection-crossroad-1016',
+#                          'person_vehicle_bike_sd512_mb2_clustered_epoch_21.pth',
+#                          True)):
+#     """ Test case for person-vehicle-bike-detection-crossroad-1016 export. """
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/tests/export_tests_person_vehicle_bike_detection.py:24]: Why is it disabled?

File: pytorch_toolkit/object_detection/tests/train_tests_face_detection.py [hunk 5]
```diff
@@ -77,29 +77,29 @@ def test_quality_metrics(self):
     return Class
 
 
-class FaceDetection0100TestCase(face_detection_test_case('face-detection-0100')):
-    """ Test case for face-detection-0100 model. """
+class FaceDetection0100TestCase(face_detection_test_case('face-detection-0200')):
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/tests/train_tests_face_detection.py:80]: Ditto.

File: pytorch_toolkit/object_detection/tests/train_tests_person_vehicle_bike_detection.py [hunk 6]
```diff
@@ -12,60 +12,60 @@
...
+#             f'python ../../external/mmdetection/tools/test.py '
+#             f'{self.configuration_file} '
+#             f'{os.path.join(self.work_dir, self.snapshot_name)} '
+#             f'--out res.pkl --eval bbox 2>&1 | tee {log_file}')
+#         ap = collect_ap(log_file)
+#
+#         with open(f'tests/expected_outputs/person-vehicle-bike-detection/{self.model_name}.json') as read_file:
+#             content = json.load(read_file)
+#
+#         self.assertEqual(content['map'], ap[0])
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/tests/train_tests_person_vehicle_bike_detection.py:71]: DItto.

File: pytorch_toolkit/object_detection/face-detection/readme.md (hunk 2 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/face-detection/readme.md:13]: @druzhkov-paul 107 and 108 won't work in mmdet2.x
@dqawami @asutic you should retrain them and send a PR again if you would like to have them in OTE. As we discussed it should not be problem for you.

File: pytorch_toolkit/object_detection/tests/train_tests_person_vehicle_bike_detection.py (hunk 6 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/tests/train_tests_person_vehicle_bike_detection.py:71]: Ditto.

File: pytorch_toolkit/object_detection/tests/train_tests_face_detection.py (hunk 5 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/tests/train_tests_face_detection.py]: fixed

File: pytorch_toolkit/object_detection/tests/export_tests_face_detection.py (hunk 3 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/tests/export_tests_face_detection.py]: fixed

File: pytorch_toolkit/object_detection/face-detection/face-detection-0200/config.py (hunk 1 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/face-detection/face-detection-0200/config.py:135]: too long to wait until first loss value appears :)

Comment (druzhkov-paul) [reply pytorch_toolkit/object_detection/face-detection/face-detection-0200/config.py:135]: Makes sense 🤣

File: pytorch_toolkit/object_detection/face-detection/readme.md (hunk 2 above)
Comment (asutic) [reply pytorch_toolkit/object_detection/face-detection/readme.md:13]: @dqawami, please go ahead and retrain the models based on the mmdet 2.0 updates.

**Comment by dqawami** [comment]
@Ilya-Krylov work_dir value for config files seems to be using the old naming convention (0100, 0102, etc.) Is there a reason for keeping it like this?

**Comment by dqawami** [comment]
@Ilya-Krylov ShuffleNet models have been updated and trained for mmdet2.0, let me know the best way to get them onto the PR

**Comment by Ilya-Krylov** [comment]
@dqawami 

> @Ilya-Krylov work_dir value for config files seems to be using the old naming convention (0100, 0102, etc.) Is there a reason for keeping it like this?

It's by mistake

You can send a PR as soon as this PR is merged. Please name your models as 0207 and 0208

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by dqawami** [comment]
@Ilya-Krylov we've decided not to include 0208 moving forward. There was an issue training a model with more than 2 out_indices using dist_train. Any insight as to why that might be happening?

**Comment by Ilya-Krylov** [comment]
@druzhkov-paul

**Comment by Ilya-Krylov** [comment]
@druzhkov-paul

**Comment by Ilya-Krylov** [comment]
run tests

File: pytorch_toolkit/object_detection/requirements.txt [hunk 7]
```diff
@@ -2,5 +2,4 @@
 -r ../../external/mmdetection/requirements.txt
 networkx==2.3
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/requirements.txt:3]: Is it necessary to have this version of `networkx`? `networkx` is in the list of MO requirements, and if you have it here only for MO, then I'd suggest to remove it.

File: pytorch_toolkit/object_detection/custom-detection/problem.yml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+description:  >-
+  Custom object detection is a computer technology being used in a variety of applications
+  that identifies human in digital images. You need to define your set of detection classes.
+favorite: false
+imagesUrls:
+  - some_picture_url
+subtitle: Detection custom objects in digital images
+title: Custom Detection
+type: generic
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/custom-detection/problem.yml:9]: It is still not clear what `Custom Detection` is. Adding a readme is a good idea. Even if this is just a project template.

File: pytorch_toolkit/object_detection/readme.md [hunk 9]
```diff
@@ -5,7 +5,10 @@
 This repo contains scripts and tutorials for object detection models training.
 
 * [Face Detection](face-detection/readme.md) - models that are able to detect faces on given images.
-* [Person Vehicle Bike Detection](person-vehicle-bike-detection/readme.md) - models that are able to detect 3 classes of objects: person, vehicle, non-vehicle (e.g. bike).
+* [Horizontal Text Detection](horizontal-text-detection/readme.md) - model that is able to detect more or less horizontal text with high speed.
+* [Person Detection](person-detection/readme.md) - models that are able to detect persons on given images.
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/readme.md]: removed

File: pytorch_toolkit/object_detection/readme.md [hunk 10]
```diff
@@ -5,7 +5,10 @@
...
+* [Vehicle Detection](vehicle-detection/readme.md) - models that are able to detect vehicles on given images.
+* [Vehicle Person Bike Detection](vehicle-person-bike-detection/readme.md) - models that are able to detect 3 classes of objects: person, vehicle, non-vehicle (e.g. bike).
 
 ## Setup
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/readme.md:13]: have set

File: pytorch_toolkit/object_detection/custom-detection/mobilenet_v2-2s_ssd-256x256/config.py [hunk 11]
```diff
@@ -89,32 +89,31 @@
...
+            img_prefix='data/coco/train2017',
             pipeline=train_pipeline
         )
     ),
     val=dict(
         type=dataset_type,
-        classes=('face',),
-        ann_file=data_root + '/val.json',
-        img_prefix=data_root,
+        # classes='TBD',
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/mobilenet_v2-2s_ssd-256x256/config.py]: removed

File: pytorch_toolkit/object_detection/custom-detection/problem.yml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+description:  >-
+  Custom object detection is a computer technology being used in a variety of applications
+  that identifies human in digital images. You need to define your set of detection classes.
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/problem.yml]: fixed

File: pytorch_toolkit/object_detection/tests/train_tests_vehicle_person_bike_detection.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+            log_file = os.path.join(self.work_dir, 'test_quality_metrics.log')
+            os.system(
+                f'python ../../external/mmdetection/tools/test.py '
+                f'{self.configuration_file} '
+                f'{os.path.join(self.work_dir, self.model_name + ".pth")} '
+                f'--out res.pkl --eval bbox 2>&1 | tee {log_file}')
+            ap = collect_ap(log_file)
+
+            with open(f'tests/expected_outputs/vehicle-person-bike-detection/{self.model_name}.json', 'w') as read_file:
+                json.dump({'map': ap[0]}, read_file)
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/tests/train_tests_vehicle_person_bike_detection.py]: 🤦 fixed

File: pytorch_toolkit/object_detection/vehicle-detection/readme.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
+# Vehicle Detection
+
+Models that are able to detect vehicles on given images.
+
+| Model Name | Complexity (GFLOPs) | Size (Mp) | AP @ [IoU=0.50:0.95] (%) | Links | GPU_NUM |
+| --- | --- | --- | --- | --- | --- |
+| vehicle-detection-0200 | 0.82 | 1.83 | 26.1 | [snapshot](https://download.01.org/opencv/openvino_training_extensions/models/object_detection/v2/vehicle-detection-0200.pth), [configuration file](./vehicle-detection-0200/config.py) | 2 |
+| vehicle-detection-0201 | 1.84 | 1.83 | 32.5 | [snapshot](https://download.01.org/opencv/openvino_training_extensions/models/object_detection/v2/vehicle-detection-0200.pth), [configuration file](./vehicle-detection-0201/config.py) | 2 |
+| vehicle-detection-0202 | 3.28 | 1.83 | 36.3 | [snapshot](https://download.01.org/opencv/openvino_training_extensions/models/object_detection/v2/vehicle-detection-0200.pth), [configuration file](./vehicle-detection-0202/config.py) | 4
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/vehicle-detection/readme.md:9]: which one?

File: pytorch_toolkit/object_detection/custom-detection/mobilenet_v2-2s_ssd-256x256/model.yml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+description: Custom Object Detection based on MobileNetV2 (SSD).
+name: mobilenet_v2-2s_ssd-256x256
+files:
+- name: snapshot.pth
+  sha256: 6632615f58746073536a96f62c60d095a5b825dbcd5d4c2d9a870cb3c6c9cf31
+  size: 8158146
+  source: https://raw.githubusercontent.com/Ilya-Krylov/tmp/master/snapshot.pth
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/mobilenet_v2-2s_ssd-256x256/model.yml]: updated

File: pytorch_toolkit/object_detection/custom-detection/problem.yml [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+favorite: false
+imagesUrls:
+  - some_picture_url
+subtitle: Detection custom objects in digital images
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/problem.yml]: fixed

File: pytorch_toolkit/object_detection/custom-detection/tools/train_and_eval.py [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+        default='')
+
+    return args.parse_args()
+
+
+def is_clustering_needed(cfg, update_config):
+    # resume_from = [p.split('=') for p in update_config.strip().split(' ') if p.startswith('resume_from=')]
+    # resume_from = resume_from[0][1] if resume_from else str(cfg.resume_from)
+    # if resume_from.lower() not in ['', 'none']:
+    #    return False
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/tools/train_and_eval.py]: removed

File: pytorch_toolkit/object_detection/custom-detection/tools/train_and_eval.py [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    subprocess.run(f'{mmdetection_tools}/dist_train.sh'
+                   f' {args.config}'
+                   f' {args.gpu_num}'
+                   f'{update_config}'.split(' '), check=True)
+
+    overrided_work_dir = [p.split('=') for p in args.update_config.strip().split(' ') if p.startswith('work_dir=')]
+    if overrided_work_dir:
+        cfg.work_dir = overrided_work_dir[0][1]
+
+    eval(args.config, os.path.join(cfg.work_dir, "latest.pth"), args.out, args.update_config)
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/tools/train_and_eval.py]: renamed

File: pytorch_toolkit/object_detection/face-detection/face-detection-0205/model.yml [hunk 19]
```diff
@@ -0,0 +1,38 @@
+description: Face Detection based on MobileNetV2 (SSD).
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/face-detection/face-detection-0205/model.yml]: fixed

File: pytorch_toolkit/object_detection/face-detection/face-detection-0206/model.yml [hunk 20]
```diff
@@ -0,0 +1,38 @@
+description: Face Detection based on MobileNetV2 (SSD).
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/face-detection/face-detection-0206/model.yml]: fixed

File: pytorch_toolkit/object_detection/requirements.txt (hunk 7 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/requirements.txt]: removed

File: pytorch_toolkit/object_detection/requirements.txt [hunk 21]
```diff
@@ -2,5 +2,4 @@
...
 opencv-python==4.1.0.25
-pycocotools==2.0.0
 tensorboardX==1.8
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/requirements.txt]: removed

File: pytorch_toolkit/object_detection/face-detection/readme.md [hunk 22]
```diff
@@ -145,7 +143,7 @@ This produces model `$MODEL_NAME.xml` and weights `$MODEL_NAME.bin` in single-pr
 For SSD networks an alternative OpenVINO™ representation is possible.
 To opt for it use extra `--alt_ssd_export` key to the `export.py` script.
 SSD model exported in such way will produce a bit different results (non-significant in most cases),
-but it also might be faster than the default one.
+but it also might be faster than the default one. Usually SSD models in [Open Model Zoo](https://github.com/opencv/open_model_zoo/) are exported using this option.
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/face-detection/readme.md]: fixed

File: pytorch_toolkit/object_detection/horizontal-text-detection/readme.md [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+*  [ICDAR2019 (MLT)](https://rrc.cvc.uab.es/?ch=15)
+*  [ICDAR2019 (ART)](https://rrc.cvc.uab.es/?ch=14) 
+*  [MSRA-TD500](http://www.iapr-tc11.org/mediawiki/index.php/MSRA_Text_Detection_500_Database_(MSRA-TD500))   
+*  [COCO-Text](https://bgshih.github.io/cocotext/)
+
+### 3. Convert datasets
+
+Extract downloaded datasets in `data/text-dataset` folder.
+
+Convert it to format that is used internally and split to the train and test part as you want.
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/horizontal-text-detection/readme.md]: fixed

File: pytorch_toolkit/object_detection/horizontal-text-detection/readme.md [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+
+     ```bash
+     python horizontal-text-detection/tools/visualize_text_detection.py \
+              $CONFIGURATION_FILE \
+              result.pkl 
+     ```
+   
+   To visualize the dependence of recall from the instance size use `--draw_graph` option.
+   
+   To show predictions on the images use `--visualize` option.
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/horizontal-text-detection/readme.md]: fixed

File: pytorch_toolkit/object_detection/horizontal-text-detection/tools/eval.py [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+def sha256sum(filename):
+    """ Computes sha256sum. """
+
+    h = hashlib.sha256()
+    b = bytearray(128*1024)
+    mv = memoryview(b)
+    with open(filename, 'rb', buffering=0) as f:
+        for n in iter(lambda: f.readinto(mv), 0):
+            h.update(mv[:n])
+    return h.hexdigest()
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/horizontal-text-detection/tools/eval.py]: fixed

File: pytorch_toolkit/object_detection/tests/train_tests_person_detection.py [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+            log_file = os.path.join(self.work_dir, 'test_quality_metrics.log')
+            os.system(
+                f'python ../../external/mmdetection/tools/test.py '
+                f'{self.configuration_file} '
+                f'{os.path.join(self.work_dir, self.model_name + ".pth")} '
+                f'--out res.pkl --eval bbox 2>&1 | tee {log_file}')
+            ap = collect_ap(log_file)
+
+            with open(f'tests/expected_outputs/person-detection/{self.model_name}.json', 'w') as read_file:
+                json.dump({'map': ap[0]}, read_file)
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/tests/train_tests_person_detection.py]: fixed

File: pytorch_toolkit/object_detection/tests/train_tests_text_detection.py [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+            log_file = os.path.join(self.work_dir, 'test_quality_metrics.log')
+            os.system(
+                f'python ../../external/mmdetection/tools/test.py '
+                f'{self.configuration_file} '
+                f'{os.path.join(self.work_dir, self.model_name + ".pth")} '
+                f'--out res.pkl --eval bbox 2>&1 | tee {log_file}')
+            ap = collect_ap(log_file)
+
+            with open(f'tests/expected_outputs/horizontal-text-detection/{self.model_name}.json', 'w') as read_file:
+                json.dump({'map': ap[0]}, read_file)
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/tests/train_tests_text_detection.py]: fixed

File: pytorch_toolkit/object_detection/tests/train_tests_person_detection.py [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+            log_file = os.path.join(self.work_dir, 'test_fine_tuning.log')
+            os.system(
+                f'../../external/mmdetection/tools/dist_train.sh {self.configuration_file} 1 2>&1 |'
+                f' tee {log_file}')
+            ap = collect_ap(log_file)
+            self.assertEqual(len((ap)), 5)
+
+        def test_quality_metrics(self):
+            log_file = os.path.join(self.work_dir, 'test_quality_metrics.log')
+            os.system(
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/tests/train_tests_person_detection.py]: fixed

File: pytorch_toolkit/object_detection/custom-detection/problem.yml [hunk 29]
```diff
@@ -0,0 +1,9 @@
+description:  >-
+  Custom object detection is a computer technology being used in a variety of applications
+  that identifies custom object in digital images. You need to define your set of detection classes.
+favorite: false
+imagesUrls:
```
Comment (Ilya-Krylov) [inline pytorch_toolkit/object_detection/custom-detection/problem.yml:5]: @dmitryagapov is it needed for custom detector in the platform? Can I remove this?

File: pytorch_toolkit/object_detection/custom-detection/problem.yml (hunk 8 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/problem.yml:9]: added

File: pytorch_toolkit/object_detection/vehicle-detection/readme.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/vehicle-detection/readme.md]: thanks, fixed

File: pytorch_toolkit/object_detection/person-detection/tools/train_and_eval.py [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def parse_args():
+    """ Parses input args. """
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('config',
+                        help='A path to model training configuration file (.py).')
+    parser.add_argument('gpu_num', type=int,
+                        help='A number of GPU to use in training.')
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/person-detection/tools/train_and_eval.py:38]:
```suggestion
                        help='A number of GPUs to use in training.')
```

File: pytorch_toolkit/object_detection/person-vehicle-bike-detection/tools/train_and_eval.py [hunk 31]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def parse_args():
+    """ Parses input args. """
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('config',
+                        help='A path to model training configuration file (.py).')
+    parser.add_argument('gpu_num', type=int,
+                        help='A number of GPU to use in training.')
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/person-vehicle-bike-detection/tools/train_and_eval.py:38]:
```suggestion
                        help='A number of GPUs to use in training.')
```

File: pytorch_toolkit/object_detection/vehicle-detection/tools/train_and_eval.py [hunk 32]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def parse_args():
+    """ Parses input args. """
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('config',
+                        help='A path to model training configuration file (.py).')
+    parser.add_argument('gpu_num', type=int,
+                        help='A number of GPU to use in training.')
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/vehicle-detection/tools/train_and_eval.py:38]:
```suggestion
                        help='A number of GPUs to use in training.')
```

File: pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0200/config.py [hunk 33]
```diff
@@ -0,0 +1,160 @@
...
+    debug=False)
+test_cfg = dict(
+    nms=dict(type='nms', iou_thr=0.45),
+    min_bbox_size=0,
+    score_thr=0.02,
+    max_per_img=200)
+# model training and testing settings
+# dataset settings
+dataset_type = 'CocoDataset'
+data_root = 'TBD'
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0200/config.py:71]: What does that mean?

File: pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0201/config.py [hunk 34]
```diff
@@ -0,0 +1,160 @@
...
+    debug=False)
+test_cfg = dict(
+    nms=dict(type='nms', iou_thr=0.45),
+    min_bbox_size=0,
+    score_thr=0.02,
+    max_per_img=200)
+# model training and testing settings
+# dataset settings
+dataset_type = 'CocoDataset'
+data_root = 'TBD'
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0201/config.py:71]: ditto

File: pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0200/config.py [hunk 35]
```diff
@@ -0,0 +1,160 @@
...
+    samples_per_gpu=185,
+    workers_per_gpu=4,
+    train=dict(
+        type='RepeatDataset',
+        times=5,
+        dataset=dict(
+            type=dataset_type,
+            classes=('vehicle',),
+            ann_file='ANNOTATION_TRAIN_TBD.json',
+            img_prefix='IMGS_PREFIX_TRAIN_TBD',
```
Comment (druzhkov-paul) [inline pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0200/config.py:115]: And here.

File: pytorch_toolkit/object_detection/custom-detection/mobilenet_v2-2s_ssd-512x512/model.yml [hunk 36]
```diff
@@ -0,0 +1,21 @@
...
+  unit: Mp
+  value: 1.99
+- display_name: Complexity
+  key: complexity
+  unit: GFLOPs
+  value: 3.42
+- display_name: AP @ [IoU=0.50:0.95]
+  key: ap
+  unit: '%'
+  value: 12.7
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/mobilenet_v2-2s_ssd-512x512/model.yml:20]: :disappointed: yes

File: pytorch_toolkit/object_detection/custom-detection/readme.md [hunk 37]
```diff
@@ -0,0 +1,135 @@
...
+export CONFIGURATION_FILE=./custom-detection/$MODEL_NAME/config.py
+```
+
+### 2. Collect dataset
+
+You can train a model on existing toy dataset `openvino_training_extensions/data/airport`. Obviously such dataset is not sufficient for training good enough model.
+
+### 3. Prepare annotation
+
+The existing toy dataset has annotation in the Common Objects in Context (COCO) and mmdetection CustomDataset format.
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/readme.md:35]: There was CustomCocoDataset that was introduced by me, but I had to remove it since the same feature (custom coco classes) appeared in CocoDataset. CustomDataset is not COCO-like format ([see here ](https://github.com/opencv/mmdetection/blob/feature/ik/more_tools_with_update_config/mmdet/datasets/custom.py#L13))

File: pytorch_toolkit/object_detection/custom-detection/mobilenet_v2-2s_ssd-256x256/config.py [hunk 38]
```diff
@@ -0,0 +1,154 @@
...
+        num_classes=80,
+        in_channels=(96, 320),
+        anchor_generator=dict(
+            type='SSDAnchorGeneratorClustered',
+            strides=(16, 32),
+            widths=[[
+                11.777124212603184, 27.156337561336, 78.40999192363739,
+                42.895380750113695
+            ],
+                    [
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/mobilenet_v2-2s_ssd-256x256/config.py]: fixed

File: pytorch_toolkit/object_detection/custom-detection/tools/train_and_eval.py [hunk 39]
```diff
@@ -0,0 +1,140 @@
...
+    update_config += f' model.bbox_head.anchor_generator.widths={str(widths).replace(" ", "")}'
+    update_config += f' model.bbox_head.anchor_generator.heights={str(heights).replace(" ", "")}'
+    logging.info('... clustering completed.')
+
+    return update_config
+
+
+def main():
+    """ Main function. """
+    logging.basicConfig(level=logging.DEBUG)
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/tools/train_and_eval.py]: changed to INFO

File: pytorch_toolkit/object_detection/face-detection/tools/eval.py [hunk 40]
```diff
@@ -31,80 +31,67 @@
...
     if not os.path.exists(eval_tools_zip):
         subprocess.run(
             f'wget http://shuoyang1213.me/WIDERFACE/support/eval_script/eval_tools.zip'
             f' -O {eval_tools_zip}'.split(' '), check=True)
-    subprocess.run(f'unzip -q -o {eval_tools_zip} -d {wider_data_folder}'.split(' '), check=True)
+    subprocess.run(f'unzip -q -o {eval_tools_zip} -d {wider_dir}'.split(' '),
+                   check=True)
 
-    wider_annotation_zip = os.path.join(wider_data_folder, 'ider_face_split.zip')
+    wider_annotation_zip = os.path.join(wider_dir, 'ider_face_split.zip')
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/face-detection/tools/eval.py]: 'w' not 'p'

File: pytorch_toolkit/object_detection/horizontal-text-detection/tools/eval.py [hunk 41]
```diff
@@ -0,0 +1,103 @@
...
+        show_dir = f' --show-dir {show_dir}' if show_dir else ''
+        subprocess.run(
+            f'python {MMDETECTION_TOOLS}/test.py'
+            f' {config_path} {snapshot}'
+            f' --out {res_pkl} --eval f1 bbox'
+            f'{show_dir}{update_config}'.split(' '), stdout=test_py_stdout,
+            check=True)
+    hmean = collect_f1(os.path.join(work_dir, 'test_py_stdout'))
+    print(hmean)
+    print(os.path.join(work_dir, 'test_py_stdout'))
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/horizontal-text-detection/tools/eval.py]: fixed

File: pytorch_toolkit/object_detection/custom-detection/readme.md [hunk 42]
```diff
@@ -0,0 +1,135 @@
...
+export NUM_CLASSES=3
+export CLASSES="vehicle,person,non-vehicle"
+export WORK_DIR="my_custom_detector"
+export UPDATE_CONFIG="model.bbox_head.num_classes=${NUM_CLASSES} \
+                      data.train.dataset.classes=${CLASSES} \
+                      data.val.classes=${CLASSES} \
+                      data.val.classes=${CLASSES} \
+                      total_epochs=20 \
+                      resume_from=${MODEL_NAME}.pth \
+                      work_dir=${WORK_DIR}"
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/readme.md:61]: added

File: pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0200/config.py (hunk 33 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0200/config.py]: There was no any dataset suitable for this task. I've added instructions how to filter `car` category from MS COCO dataset and updated corresponding fields in configuration files.

File: pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0201/config.py (hunk 34 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0201/config.py]: ditto

File: pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0200/config.py (hunk 35 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/vehicle-detection/vehicle-detection-0200/config.py]: And  here.

File: pytorch_toolkit/object_detection/custom-detection/problem.yml (hunk 29 above)
Comment (Ilya-Krylov) [reply pytorch_toolkit/object_detection/custom-detection/problem.yml]: It is not used. Removed

**Comment by Ilya-Krylov** [comment]
@druzhkov-paul all comments seems to be resolved
